File: src/api/playlists.ts
import { invoke } from "@tauri-apps/api/core";
import type { Playlist, Track } from "../types";

export async function getPlaylists(): Promise<Playlist[]> {
  return invoke("get_playlists");
}

export async function getPlaylistTracks(
  playlistId: number,
  limit?: number,
  offset?: number,
): Promise<Track[]> {
  return invoke("get_playlist_tracks", { playlistId, limit, offset });
}

export async function createPlaylist(
  name: string,
  parentPersistentId?: string | null,
  isFolder?: boolean,
): Promise<Playlist> {
  return invoke("create_playlist", {
    name,
    parentPersistentId: parentPersistentId ?? null,
    isFolder: isFolder ?? false,
  });
}

export async function renamePlaylist(
  playlistId: number,
  name: string,
): Promise<void> {
  return invoke("rename_playlist", { playlistId, name });
}

export async function deletePlaylist(playlistId: number): Promise<void> {
  return invoke("delete_playlist", { playlistId });
}

export async function addTracksToPlaylist(
  playlistId: number,
  trackIds: number[],
): Promise<number> {
  return invoke("add_tracks_to_playlist", { playlistId, trackIds });
}

export async function removeTrackFromPlaylist(
  playlistId: number,
  sortIndex: number,
): Promise<void> {
  return invoke("remove_track_from_playlist", { playlistId, sortIndex });
}

export async function reorderPlaylistTracks(
  playlistId: number,
  orderedTrackIds: number[],
): Promise<void> {
  return invoke("reorder_playlist_tracks", { playlistId, orderedTrackIds });
}

File: src-tauri/src/commands/playlists.rs
use tauri::AppHandle;

use crate::commands::library::open_db;
use crate::models::{Playlist, Track};

#[tauri::command]
pub fn get_playlists(app: AppHandle) -> Result<Vec<Playlist>, String> {
    let db = open_db(&app)?;
    db.get_playlists().map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_playlist_tracks(
    app: AppHandle,
    playlist_id: i64,
    limit: Option<i64>,
    offset: Option<i64>,
) -> Result<Vec<Track>, String> {
    let db = open_db(&app)?;
    db.get_playlist_tracks(playlist_id, limit.unwrap_or(500), offset.unwrap_or(0))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn create_playlist(
    app: AppHandle,
    name: String,
    parent_persistent_id: Option<String>,
    is_folder: Option<bool>,
) -> Result<Playlist, String> {
    let db = open_db(&app)?;
    db.create_playlist(
        &name,
        parent_persistent_id.as_deref(),
        is_folder.unwrap_or(false),
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn rename_playlist(app: AppHandle, playlist_id: i64, name: String) -> Result<(), String> {
    let db = open_db(&app)?;
    db.rename_playlist(playlist_id, &name)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_playlist(app: AppHandle, playlist_id: i64) -> Result<(), String> {
    let db = open_db(&app)?;
    db.delete_playlist(playlist_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn add_tracks_to_playlist(
    app: AppHandle,
    playlist_id: i64,
    track_ids: Vec<i64>,
) -> Result<usize, String> {
    let db = open_db(&app)?;
    db.add_tracks_to_playlist(playlist_id, &track_ids)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn remove_track_from_playlist(
    app: AppHandle,
    playlist_id: i64,
    sort_index: i64,
) -> Result<(), String> {
    let db = open_db(&app)?;
    db.remove_track_from_playlist(playlist_id, sort_index)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn reorder_playlist_tracks(
    app: AppHandle,
    playlist_id: i64,
    ordered_track_ids: Vec<i64>,
) -> Result<(), String> {
    let db = open_db(&app)?;
    db.reorder_playlist_tracks(playlist_id, &ordered_track_ids)
        .map_err(|e| e.to_string())
}
